gtsam/discrete/DiscreteFactorGraph.cpp

```diff
@@ -65,11 +65,18 @@ namespace gtsam {
 
   /* ************************************************************************ */
   DecisionTreeFactor DiscreteFactorGraph::product() const {
-    DecisionTreeFactor result;
-    for (const sharedFactor& factor : *this) {
-      if (factor) result = (*factor) * result;
+    DiscreteFactor::shared_ptr result;
+    for (auto it = this->begin(); it != this->end(); ++it) {
```

Comment:
Why is it not just
```
for (const sharedFactor& factor : *this) 
      if (factor) result = factor-> multiply(result);
```

Reply:
Because `result` is initially null. So `factor*result` will give null every time.

Reply:
Reminder that `result` is now a shared_ptr not an object, so it initially points to nothing and has no data associated with it.